File: training-data/swift/template_method_pattern.swift
protocol DataExporter {
    func fetchData() -> [String]
    func format(_ rows: [String]) -> String
    func header() -> String
}

extension DataExporter {
    func export() -> String {
        let rows = fetchData()
        return header() + "\n" + format(rows)
    }
}

struct CSVExporter: DataExporter {
    func fetchData() -> [String] { ["a,1", "b,2", "c,3"] }
    func header() -> String { "name,value" }
    func format(_ rows: [String]) -> String { rows.joined(separator: "\n") }
}

struct JSONExporter: DataExporter {
    func fetchData() -> [String] { ["a", "b", "c"] }
    func header() -> String { "// generated export" }
    func format(_ rows: [String]) -> String {
        "[" + rows.map { "\"\($0)\"" }.joined(separator: ",") + "]"
    }
}

let exporters: [DataExporter] = [CSVExporter(), JSONExporter()]
for exporter in exporters {
    print(exporter.export())
    print("---")
}

File: training-data/php/template_method_pattern.php
<?php

abstract class DataExporter
{
    final public function export(array $rows): string
    {
        $header = $this->formatHeader();
        $body = implode($this->rowSeparator(), array_map(fn ($row) => $this->formatRow($row), $rows));
        return $header . $body;
    }

    abstract protected function formatHeader(): string;
    abstract protected function formatRow(array $row): string;

    protected function rowSeparator(): string
    {
        return "\n";
    }
}

class CsvExporter extends DataExporter
{
    protected function formatHeader(): string
    {
        return "name,age\n";
    }

    protected function formatRow(array $row): string
    {
        return "{$row['name']},{$row['age']}";
    }
}

class JsonLinesExporter extends DataExporter
{
    protected function formatHeader(): string
    {
        return '';
    }

    protected function formatRow(array $row): string
    {
        return json_encode($row);
    }
}

$rows = [['name' => 'Alice', 'age' => 30], ['name' => 'Bob', 'age' => 25]];
echo (new CsvExporter())->export($rows) . "\n";
echo (new JsonLinesExporter())->export($rows) . "\n";
